awspub/image.py

```diff
@@ -500,8 +459,71 @@ def create(self) -> Dict[str, _ImageInfo]:
         if self.conf["share"]:
             self._share(self.conf["share"], images)
 
+        if missing_regions:
+            logger.error("Failed to publish snapshot to all regions", extra={"missing_regions": missing_regions})
+            raise exceptions.IncompleteSnapshotSetException("Incomplete snapshot set published")
+
         return images
 
+    def _register_snapshots(self, snapshot_ids: Dict[str, str], region: str, ec2client_region: EC2Client):
+        logger.info(
+            f"creating image with name '{self.image_name}' in " f"region {ec2client_region.meta.region_name} ..."
+        )
+
+        register_image_kwargs = dict(
+            Name=self.image_name,
+            Description=self.conf.get("description", ""),
+            Architecture=self._ctx.conf["source"]["architecture"],
+            RootDeviceName=self.conf["root_device_name"],
+            BlockDeviceMappings=[
+                {
+                    "Ebs": {
+                        "SnapshotId": snapshot_ids[region],
+                        "VolumeType": self.conf["root_device_volume_type"],
+                        "VolumeSize": self.conf["root_device_volume_size"],
+                    },
+                    "DeviceName": self.conf["root_device_name"],
+                },
+                # TODO: make those ephemeral block device mappings configurable
+                {"VirtualName": "ephemeral0", "DeviceName": "/dev/sdb"},
+                {"VirtualName": "ephemeral1", "DeviceName": "/dev/sdc"},
+            ],
+            EnaSupport=True,
+            SriovNetSupport="simple",
+            VirtualizationType="hvm",
+            BootMode=self.conf["boot_mode"],
+        )
+
+        if self.conf["tpm_support"]:
+            register_image_kwargs["TpmSupport"] = self.conf["tpm_support"]
+
+        if self.conf["imds_support"]:
+            register_image_kwargs["ImdsSupport"] = self.conf["imds_support"]
+
+        if self.conf["uefi_data"]:
+            with open(self.conf["uefi_data"], "r") as f:
+                uefi_data = f.read()
+            register_image_kwargs["UefiData"] = uefi_data
+
+        if self.conf["billing_products"]:
+            register_image_kwargs["BillingProducts"] = self.conf["billing_products"]
+
+        try:
+            resp = ec2client_region.register_image(**register_image_kwargs)
+            ec2client_region.create_tags(Resources=[resp["ImageId"]], Tags=self._tags)
+        except botocore.exceptions.ClientError as e:
```

Comment:
can you move that up so `create_tags` is outside of this exception? otherwise if `create_tags` does raise a `OperationNotPermitted` exception, the error message would be very wrong.